notifications/manager.go

```diff
@@ -0,0 +1,205 @@
+	err := n.registerNotifications(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// registerNotifications registers a new server notification stream.
+func (n *Manager) registerNotifications(mainCtx context.Context) error {
+	// In order to create a valid l402 we first are going to call
+	// the FetchL402 method. As a client might not have outbound capacity
+	// yet, we'll retry until we get a valid response.
+	if !n.hasL402 {
+		n.fetchL402(mainCtx)
+	}
+
+	callCtx, cancel := context.WithCancel(mainCtx)
+
+	// We'll now subscribe to the swap server notifications.
+	notifStream, err := n.cfg.Client.SubscribeNotifications(
+		callCtx, &swapserverrpc.SubscribeNotificationsRequest{},
+	)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	log.Debugf("Successfully subscribed to server notifications")
+
+	// We'll now start a goroutine that will handle all the incoming
+	// notifications.
+	go func() {
+		for {
+			notification, err := notifStream.Recv()
+			if err == nil && notification != nil {
+				log.Debugf("Received notification: %v", notification)
+				n.handleNotification(notification)
+				continue
+			}
+			log.Errorf("Error receiving "+
+				"notification: %v", err)
+
+			cancel()
+
+			// If we encounter an error, we'll
+			// try to reconnect.
+			for {
+				select {
+				case <-mainCtx.Done():
+					return
+
+				case <-time.After(time.Second * 10):
+					log.Debugf("Reconnecting to server notifications")
+					err = n.registerNotifications(mainCtx)
+					if err != nil {
+						log.Errorf("Error reconnecting: %v", err)
+						continue
+					}
+
+					// If we were able to reconnect, we'll
+					// return.
+					return
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+// fetchL402 fetches the L402 from the server. This method will keep on
+// retrying until it gets a valid response.
+func (m *Manager) fetchL402(ctx context.Context) {
+	// Add a 0 timer so that we initially fetch the L402 immediately.
+	timer := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case <-timer.C:
+			err := m.cfg.FetchL402(ctx)
+			if err != nil {
+				log.Warnf("Error fetching L402: %v", err)
+				timer.Reset(time.Second * 10)
+				continue
+			}
+			m.hasL402 = true
+
+			return
+		}
+	}
+}
+
+// handleNotification handles an incoming notification from the server,
+// forwarding it to the appropriate subscribers.
+func (n *Manager) handleNotification(notification *swapserverrpc.
+	SubscribeNotificationsResponse) {
+
+	switch notification.Notification.(type) {
+	case *swapserverrpc.SubscribeNotificationsResponse_ReservationNotification:
+		// We'll forward the reservation notification to all subscribers.
+		// Cleaning up any subscribers that have been canceled.
+		newSubs := []subscriber{}
+		for _, sub := range n.subscribers[NotificationTypeReservation] {
+			recvChan := sub.recvChan.(chan *swapserverrpc.
+				ServerReservationNotification)
+
+			select {
+			case <-sub.subCtx.Done():
+				continue
```

Comment:
I propose to close recvChan here.

Note that it will be closed not right after ctx expires, but after first notification of this type is received. This is fine, because receiving side (the caller) is likely to wait on the same ctx expiry anyway in `select` statement.